node/external/transactionAPI/unmarshalAndPrepare.go

```diff
@@ -0,0 +1,129 @@
+package transactionAPI
+
+import (
+	"encoding/hex"
+	"github.com/ElrondNetwork/elrond-go-core/core"
+	rewardTxData "github.com/ElrondNetwork/elrond-go-core/data/rewardTx"
+	"github.com/ElrondNetwork/elrond-go-core/data/smartContractResult"
+	"github.com/ElrondNetwork/elrond-go-core/data/transaction"
+	"github.com/ElrondNetwork/elrond-go-core/marshal"
+)
+
+type unmarshalerAndPreparer struct {
+	addressPubKeyConverter core.PubkeyConverter
+	marshalizer            marshal.Marshalizer
+}
+
+func newTransactionUnmashalerAndPreparer(marshalizer marshal.Marshalizer, addressPubKeyConverter core.PubkeyConverter) *unmarshalerAndPreparer {
+	return &unmarshalerAndPreparer{
+		marshalizer:            marshalizer,
+		addressPubKeyConverter: addressPubKeyConverter,
+	}
+}
+
+func (up *unmarshalerAndPreparer) unmarshalTransaction(txBytes []byte, txType transaction.TxType) (*transaction.ApiTransactionResult, error) {
+	switch txType {
+	case transaction.TxTypeNormal:
+		var tx transaction.Transaction
+		err := up.marshalizer.Unmarshal(&tx, txBytes)
+		if err != nil {
+			return nil, err
+		}
+		return up.prepareNormalTx(&tx)
+	case transaction.TxTypeInvalid:
+		var tx transaction.Transaction
+		err := up.marshalizer.Unmarshal(&tx, txBytes)
+		if err != nil {
+			return nil, err
+		}
+		return up.prepareInvalidTx(&tx)
+	case transaction.TxTypeReward:
+		var tx rewardTxData.RewardTx
+		err := up.marshalizer.Unmarshal(&tx, txBytes)
+		if err != nil {
+			return nil, err
+		}
+		return up.prepareRewardTx(&tx)
+
+	case transaction.TxTypeUnsigned:
+		var tx smartContractResult.SmartContractResult
+		err := up.marshalizer.Unmarshal(&tx, txBytes)
+		if err != nil {
+			return nil, err
+		}
+		return up.prepareUnsignedTx(&tx)
+	}
+
+	return &transaction.ApiTransactionResult{Type: string(transaction.TxTypeInvalid)}, nil // this shouldn't happen
```

Comment:
Should remove the comment: it is misleading. It can happen if we add a new type of transaction handler and forget to add it in this struct

Reply:
done.